### PR TITLE
Support native namespace functions with type dispatch.

### PR DESCRIPTION
**Pull request by gchanan**
Use 'ones' as an example.  Note this is a "halfway" solution; i.e. the call chain is:
at::ones(shape, dtype) -> dtype.ones(shape, dtype) -> CPUFloatType.ones(shape, dtype) -> at::native::ones(shape, dtype)

The "nicer" solution would be:
at::ones(shape, dtype) -> dtype.ones(shape) -> CPUFloatType.ones(shape) -> at::native::ones(shape, this)

### CODE REVIEW

**Comment by gchanan** [comment]
The downside of this approach compared to the "nicer" solution:
1) It breaks existing code that calls directly into the Type (and makes calls like CPUFloatType.ones({}, CPUDoubleType) weird, but this seems okay because the Type object is really a dispatch implementation detail, and we should prefer that people use the namespace function.
2) You wouldn't hand-write things this way

The upside is it avoids complicating the dispatch; i.e. since the Type methods always match the function methods, a lot more code generation doesn't need to change.  I implemented the other one, but I wasn't happy with how complex things were getting.

**Review by colesbury — APPROVED** [review]

**Comment by ezyang** [comment]
ASAN log:

```
18:46:44 =================================================================
18:46:44 ==411==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f73fdf55d85 bp 0x7fffd6bf5b60 sp 0x7fffd6bf5880 T0)
18:46:44 ==411==The signal is caused by a WRITE memory access.
18:46:44 ==411==Hint: address points to the zero page.
18:46:45     #0 0x7f73fdf55d84 in at::native::randperm_out(at::Tensor&, long, at::Generator*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libATen.so.1+0x1e3d84)
18:46:45     #1 0x7f73fe4ece4e in at::Type::randperm_out(at::Tensor&, long, at::Generator*) const (/opt/conda/lib/python3.6/site-packages/torch/lib/libATen.so.1+0x77ae4e)
18:46:46     #2 0x7f7400faf1cd in torch::autograd::VariableType::randperm_out(at::Tensor&, long, at::Generator*) const /var/lib/jenkins/workspace/torch/csrc/autograd/generated/VariableType.cpp:19338:9
18:46:46     #3 0x7f7401416e84 in at::randperm_out(at::Tensor&, long, at::Generator*) /var/lib/jenkins/workspace/torch/lib/tmp_install/include/ATen/Functions.h:3035:31
18:46:46     #4 0x7f7401416e84 in torch::autograd::dispatch_randperm(long, at::Generator*, at::Tensor) /var/lib/jenkins/workspace/torch/csrc/autograd/generated/python_torch_functions_dispatch.h:2061
18:46:46     #5 0x7f7401416e84 in torch::autograd::THPVariable_randperm(_object*, _object*, _object*) /var/lib/jenkins/workspace/torch/csrc/autograd/generated/python_torch_functions.cpp:3270
18:46:46     #6 0x55832c44cfd3 in _PyCFunction_FastCallDict (/opt/conda/bin/python3.6+0x10efd3)
18:46:46     #7 0x55832c47af23 in _PyCFunction_FastCallKeywords (/opt/conda/bin/python3.6+0x13cf23)
18:46:46     #8 0x55832c4dabeb in call_function (/opt/conda/bin/python3.6+0x19cbeb)
18:46:46     #9 0x55832c4ffeb0 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1eb0)
18:46:46     #10 0x55832c4d47da in fast_function (/opt/conda/bin/python3.6+0x1967da)
18:46:46     #11 0x55832c4dacc4 in call_function (/opt/conda/bin/python3.6+0x19ccc4)
18:46:46     #12 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #13 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #14 0x55832c4d5107 in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x197107)
18:46:46     #15 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #16 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #17 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #18 0x55832c500993 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c2993)
18:46:46     #19 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #20 0x55832c4d4eed in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x196eed)
18:46:46     #21 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #22 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #23 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #24 0x55832c4a8900 in slot_tp_call (/opt/conda/bin/python3.6+0x16a900)
18:46:46     #25 0x55832c44d1ba in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f1ba)
18:46:46     #26 0x55832c4dad3d in call_function (/opt/conda/bin/python3.6+0x19cd3d)
18:46:46     #27 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #28 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #29 0x55832c4d5107 in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x197107)
18:46:46     #30 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #31 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #32 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #33 0x55832c500993 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c2993)
18:46:46     #34 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #35 0x55832c4d4eed in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x196eed)
18:46:46     #36 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #37 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #38 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #39 0x55832c4a8900 in slot_tp_call (/opt/conda/bin/python3.6+0x16a900)
18:46:46     #40 0x55832c44d1ba in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f1ba)
18:46:46     #41 0x55832c4dad3d in call_function (/opt/conda/bin/python3.6+0x19cd3d)
18:46:46     #42 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #43 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #44 0x55832c4d5107 in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x197107)
18:46:46     #45 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #46 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #47 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #48 0x55832c500993 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c2993)
18:46:46     #49 0x55832c4d39a5 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x1959a5)
18:46:46     #50 0x55832c4d4eed in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x196eed)
18:46:46     #51 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #52 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #53 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #54 0x55832c4a8900 in slot_tp_call (/opt/conda/bin/python3.6+0x16a900)
18:46:46     #55 0x55832c44d1ba in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f1ba)
18:46:46     #56 0x55832c4dad3d in call_function (/opt/conda/bin/python3.6+0x19cd3d)
18:46:46     #57 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #58 0x55832c4d47da in fast_function (/opt/conda/bin/python3.6+0x1967da)
18:46:46     #59 0x55832c4dacc4 in call_function (/opt/conda/bin/python3.6+0x19ccc4)
18:46:46     #60 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #61 0x55832c4d47da in fast_function (/opt/conda/bin/python3.6+0x1967da)
18:46:46     #62 0x55832c4dacc4 in call_function (/opt/conda/bin/python3.6+0x19ccc4)
18:46:46     #63 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #64 0x55832c4d3f90 in _PyEval_EvalCodeWithName (/opt/conda/bin/python3.6+0x195f90)
18:46:46     #65 0x55832c4d5107 in _PyFunction_FastCallDict (/opt/conda/bin/python3.6+0x197107)
18:46:46     #66 0x55832c44d39e in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f39e)
18:46:46     #67 0x55832c451ff2 in _PyObject_Call_Prepend (/opt/conda/bin/python3.6+0x113ff2)
18:46:46     #68 0x55832c44cddd in PyObject_Call (/opt/conda/bin/python3.6+0x10eddd)
18:46:46     #69 0x55832c4a7f6a in slot_tp_init (/opt/conda/bin/python3.6+0x169f6a)
18:46:46     #70 0x55832c4daf26 in type_call (/opt/conda/bin/python3.6+0x19cf26)
18:46:46     #71 0x55832c44d1ba in _PyObject_FastCallDict (/opt/conda/bin/python3.6+0x10f1ba)
18:46:46     #72 0x55832c4d4cf9 in _PyObject_FastCallKeywords (/opt/conda/bin/python3.6+0x196cf9)
18:46:46     #73 0x55832c4dad3d in call_function (/opt/conda/bin/python3.6+0x19cd3d)
18:46:46     #74 0x55832c4ffeb0 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1eb0)
18:46:46     #75 0x55832c4d47da in fast_function (/opt/conda/bin/python3.6+0x1967da)
18:46:46     #76 0x55832c4dacc4 in call_function (/opt/conda/bin/python3.6+0x19ccc4)
18:46:46     #77 0x55832c4ff199 in _PyEval_EvalFrameDefault (/opt/conda/bin/python3.6+0x1c1199)
18:46:46     #78 0x55832c4d5528 in PyEval_EvalCodeEx (/opt/conda/bin/python3.6+0x197528)
18:46:46     #79 0x55832c4d62cb in PyEval_EvalCode (/opt/conda/bin/python3.6+0x1982cb)
18:46:46     #80 0x55832c552af3 in run_mod (/opt/conda/bin/python3.6+0x214af3)
18:46:46     #81 0x55832c552ef0 in PyRun_FileExFlags (/opt/conda/bin/python3.6+0x214ef0)
18:46:46     #82 0x55832c5530f3 in PyRun_SimpleFileExFlags (/opt/conda/bin/python3.6+0x2150f3)
18:46:46     #83 0x55832c556c27 in Py_Main (/opt/conda/bin/python3.6+0x218c27)
18:46:46     #84 0x55832c41e71d in main (/opt/conda/bin/python3.6+0xe071d)
18:46:46     #85 0x7f740af0c82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
18:46:46     #86 0x55832c505c97 in _start (/opt/conda/bin/python3.6+0x1c7c97)
18:46:46 
18:46:46 AddressSanitizer can not provide additional info.
18:46:46 SUMMARY: AddressSanitizer: SEGV (/opt/conda/lib/python3.6/site-packages/torch/lib/libATen.so.1+0x1e3d84) in at::native::randperm_out(at::Tensor&, long, at::Generator*)
```

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+
+Tensor& arange_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+  return at::_arange_out(result, start, end, step);
+}
+
+Tensor arange(const Type&dtype, Scalar end) {
+  return dtype._arange(end);
+}
+
+Tensor& arange_out(Tensor &result, Scalar end) {
```
Comment (colesbury) [inline aten/src/ATen/native/TensorFactories.cpp]: `Tensor& result` to be consistent with the rest of the file

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -12,28 +34,178 @@ Tensor empty_like(const Tensor& self, const Type& dtype) {
...
+Tensor eye(const Type& dtype, int64_t n, int64_t m) {
+  auto result = dtype.tensor();
+  return at::eye_out(result, n, m);
+}
+
+Tensor& eye_out_cpu(Tensor& result, int64_t n, int64_t m) {
+  if (n <= 0) {
+    std::ostringstream oss;
+    oss << "n must be greater than 0, got: " << n;
+    std::runtime_error(oss.str());
```
Comment (colesbury) [inline aten/src/ATen/native/TensorFactories.cpp]: `throw std::runtime_error`

**Review by colesbury — APPROVED** [review]
lgtm

**Comment by gchanan** [comment]
This now implements the "nicer" solution described above, with the different that the Type is now the first argument (although that is not a strict requirement) to match the _out variants.  So the call order is now:
at::ones(dtype, shape) -> dtype.ones(shape) -> CPUFloatType.ones(shape) -> at::native::ones(*this, shape)

**Pull request merged**